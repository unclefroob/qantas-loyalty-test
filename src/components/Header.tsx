import { Box, styled, Typography } from "@mui/material";
import { ChangeEvent, Dispatch, SetStateAction } from "react";

export enum SortBy {
  DESC = "DESC",
  ASC = "ASC",
}
interface IHeaderProps {
  numberOfHotels: number;
  sortBy: SortBy;
  setSortBy: Dispatch<SetStateAction<SortBy>>;
}

const HeaderBox = styled(Box)({
  display: "flex",
  justifyContent: "space-between",
  padding: "0.4rem",
});

const SortByContainer = styled(Box)({
  display: "flex",
});

const Header = (props: IHeaderProps) => {
  const { numberOfHotels, sortBy, setSortBy } = props;

  const handleSort = (e: ChangeEvent<HTMLSelectElement>) => {
    setSortBy(SortBy[e.target.value as keyof typeof SortBy]);
  };

  return (
    <HeaderBox>
      <Typography>{numberOfHotels} hotels in Sydney</Typography>
      <SortByContainer>
        <Typography sx={{ marginRight: "0.2rem" }}>Sort by</Typography>
        <select onChange={handleSort} value={sortBy}>
          <option value={SortBy.DESC}>Price (High to Low)</option>
          <option value={SortBy.ASC}>Price (Low to High)</option>
        </select>
      </SortByContainer>
    </HeaderBox>
  );
};

export default Header;
