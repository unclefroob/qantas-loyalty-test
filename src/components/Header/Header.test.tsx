import { describe, it, vi, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import Header, { SortBy } from "./Header";
import userEvent from "@testing-library/user-event";
import { useState } from "react";

const mockSetSortBy = vi.fn();

describe("Header", () => {
  it("renders the Header component", () => {
    const { asFragment } = render(
      <Header
        numberOfHotels={1}
        sortBy={SortBy.PRICE_ASC}
        setSortBy={mockSetSortBy}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("calling onChange when the sortby menu value is changed", async () => {
    const MockHeader = () => {
      const [sortBy, setSortBy] = useState(SortBy.PRICE_ASC);
      return (
        <Header numberOfHotels={1} sortBy={sortBy} setSortBy={setSortBy} />
      );
    };
    render(<MockHeader />);
    const menu = screen.getByTestId("sort-menu");
    expect(menu).toHaveValue(SortBy.PRICE_ASC);
    await userEvent.selectOptions(menu, SortBy.PRICE_DESC);
    expect(menu).toHaveValue(SortBy.PRICE_DESC);
  });
});
