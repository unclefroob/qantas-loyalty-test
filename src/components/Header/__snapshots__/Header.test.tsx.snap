// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > renders the Header component 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-vf7s70"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
    >
      1 hotels in Sydney
    </p>
    <div
      class="MuiBox-root css-162tvoi"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1v4yp5n-MuiTypography-root"
      >
        Sort by
      </p>
      <select
        data-testid="sort-menu"
      >
        <option
          value="PRICE_DESC"
        >
          Price (High to Low)
        </option>
        <option
          value="PRICE_ASC"
        >
          Price (Low to High)
        </option>
      </select>
    </div>
  </div>
</DocumentFragment>
`;
